src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -553,35 +553,52 @@ class Wavefunction
     /// \pre: Each qubit, listed in `q`, must be unentangled and in state |0>.
     /// Place qubits, listed in `q` into superposition of basis vectors with provided `amplitudes`, where the order of
     /// qubits in array `q` defines the order of the basis vectors (little endian).
-    /// NB: at the moment supported only total state injection on all currently allocated qubits
     void inject_state(const std::vector<logical_qubit_id>& qubits, const std::vector<ComplexType>& amplitudes)
     {
         assert((static_cast<size_t>(1) << qubits.size()) == amplitudes.size());
 
-        if (num_qubits_ != qubits.size()) throw std::runtime_error("Subsystem state injection not supported (yet)");
-
         flush();
 
         // Check prerequisites.
         for (logical_qubit_id q : qubits)
         {
-            if (!kernels::isclassical(wfn_, get_qubit_position(q))
-                || kernels::getvalue(wfn_, get_qubit_position(q)) != 0)
+            if (!kernels::isclassical(wfn_, get_qubit_position(q)) ||
+                kernels::getvalue(wfn_, get_qubit_position(q)) != 0)
             {
                 throw std::runtime_error("Cannot prepare state of entangled qubits or if they are not in state |0>");
             }
         }
 
-        // Reorder the qubits to the positions that match the wave function provided by the user.
-        for (unsigned i = 0; i < qubits.size(); i++)
+        if (qubits.size() == num_qubits_)
         {
-            qubitmap_[qubits[i]] = i;
+            // For full state injection we can simply copy the user's wave function into our store and reorder the
+            // positions map.
+            for (unsigned i = 0; i < qubits.size(); i++)
+            {
+                qubitmap_[qubits[i]] = i;
+            }
+            wfn_.assign(amplitudes.begin(), amplitudes.end());
         }
-
-        // For full state injection we can simply copy the user's wave function into our store.
-        for (size_t i = 0; i < wfn_.size(); i++)
+        else
         {
-            wfn_[i] = amplitudes[i];
+            const int64_t num_states = static_cast<int64_t>(wfn_.size());
+            std::vector<positional_qubit_id> positions = get_qubit_positions(qubits);
```

Comment:
This is a very minor optimization, but now this branch ends up iterating over `qubits` to convert them into positional ids, and the precheck on each entry in `qubits` above iterates over the whole list too. Perhaps this positional vector should be created at the top of the function and used in both the precheck and here?

Reply:
This would be a trade-off, I suppose, and I don't know which is better: the cost of always allocating the vector to save one additional loop through the `qubits` input, or only allocating the positional vector when we need it for subsystem injection at the cost of iterating through `qubits` twice. I don't expect `qubits` to be very big, so I'm not sure which side to err on.

Reply:
I honestly don't think it matters here because, as you said, qubits array isn't going to be large, but getting the positions vector before doing the prereq loop will make the loop a bit cleaner looking, so why not? :)